### DISCUSSION TITLE
Why Next.js raise error 502 Bad Gateway when loading / rewriting image from AWS S3?

### DISCUSSION THREAD

**Discussion by j4nos** [vercel/vercel#8168] [Help] [answered]
We use middleware for url rewrite, and when image is loaded from AWS S3, Next.js / Vercel raise this error:

ROUTER_EXTERNAL_TARGET_HANDSHAKE_ERROR

when opening: 

https://tikex-dev.com/kubl/b1a1/j1vd/vree

Middleware should load in this image: https://t44-post-cover.s3.eu-central-1.amazonaws.com/vree-patch and display it.

```jsx
//middleware.ts

import { NextResponse, userAgent } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
	const hostname = request.headers.get('host')
	const { ua } = userAgent(request)
	//ua.includes('facebook') || hostname.includes('facebook.com')
	if (true) {
		const url = request.nextUrl.clone()
		const a = url.pathname.split('/')
		url.hostname = 't44-post-cover.s3.eu-central-1.amazonaws.com'
		url.pathname = a[a.length - 1] + '-patch'
		url.port = '80'
		return NextResponse.rewrite(url)
	}
}

export const config = {
	matcher: ['/:organizationShortId/:postId/:tempUserShortId/:imgId'],
}
```

Following this tutorial: https://www.ctnicholas.dev/articles/how-to-use-nextjs-middleware

[![enter image description here][1]][1]

  [1]: https://i.stack.imgur.com/1AbrU.png

---

Is it possible Vercel can not handle a filesize of 6MB?

**Comment by amyegan** [comment, marked as answer]
Hi @j4nos. It looks like you may have already worked it out. I'm not seeing the error on your example page anymore. 

I'm going to include a couple of ideas here anyway, just in case they help in the future. 
- [Edge Middleware](https://vercel.com/docs/concepts/functions/edge-middleware) has a [memory limit of 128MiB](https://vercel.com/docs/concepts/functions/edge-middleware/limitations#memory), and exceeding this will cause a `502` error. 
- We have support articles with some suggestions for [debugging 502 errors](https://vercel.com/support/articles/how-to-debug-a-502-error) and [reducing serverless function size](https://vercel.com/support/articles/how-to-bypass-vercel-body-size-limit-serverless-functions). I know those are about serverless functions rather than middleware, but I think some of the concepts there may be helpful for this issue as well.
